### PR TITLE
Implement Support for Express Model

### PR DESCRIPTION
**Pull request by c-mueller**
# ToDo

This ToDo List is not only about this PR it also contains tasks way too large for this PR

- [x] Implement Dynamic Port Detection for WebApplications/Endpoint Capabilites
- [ ] Support Building of more abstract types Like DBMS and SoftwareComponent
    - [ ] SoftwareComponent
    - [x] DBMS
    - [ ] WebServer
- [ ] Implement Support for DockerApplication Nodes
- [ ] Deployment Order (Considered irrelevant for now! *Why?* If a Pod fails (errors with code `!= 0` Kubernetes will attempt to restart it. This represents a loop waiting for the dependencies to be created)
    - [ ] Deploy `ConnectsTo` sources after the target
    - [ ] Implement Similar behaviour for `DependsOn`
- [ ] Investigate Transformation of Mutli Application -> Webserver and Multi Database -> DBMS
- [ ] Handle Scalability (Based on the ScalableCapability of a Compute Node)
- [ ] Refactor Kubernetes Plugin
- [ ] Think about Persitency (Block Storage)

### CODE REVIEW

File: server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/KubernetesLampIT.java [hunk 1]
```diff
@@ -24,17 +28,26 @@
...
+
     @Override
     protected EffectiveModel getModel() {
         return new EffectiveModelFactory().create(TestCsars.VALID_LAMP_NO_INPUT_TEMPLATE, logMock());
     }
 
     @Override
-    protected void onSuccess(File outputDir) {
+    protected void onSuccess(File outputDir) throws InterruptedException {
+        Thread.sleep(10000);
```
Comment (hnicke) [inline server/src/integration/java/org/opentosca/toscana/plugins/kubernetes/KubernetesLampIT.java]: I guess this was just for debugging? Can this get removed?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java [hunk 2]
```diff
@@ -69,7 +69,7 @@ public void progress(ProgressMessage progressMessage) throws DockerException {
             log(progressMessage.progress(), logger::trace);
         log(progressMessage.stream(), logger::info);
         log(progressMessage.error(), logger::error);
-        log(progressMessage.status(), logger::info);
+        log(progressMessage.status(), logger::trace);
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java]: Can this get set back to `info`?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/visitor/imgtransform/DockerfileBuildingVisitor.java [hunk 3]
```diff
@@ -74,12 +100,22 @@ public void visit(Apache node) {
 
     @Override
     public void visit(WebApplication node) {
+        //If a webapplication does not define a port. add the defaults for HTTP and HTTPS
+        if (!node.getAppEndpoint().getPort().isPresent()) {
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/visitor/imgtransform/DockerfileBuildingVisitor.java]: Wouldn't this be the perfect moment to set the port in the model as well?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/visitor/imgtransform/DockerfileBuildingVisitor.java [hunk 4]
```diff
@@ -74,12 +100,22 @@ public void visit(Apache node) {
...
     }
 
     @Override
     public void visit(MysqlDbms node) {
-        ports.add(3306);
+        if (node.getPort().isPresent()) {
+            ports.add(node.getPort().get());
+        } else {
+            //Add mysql default port if none is set
+            ports.add(3306);
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/visitor/imgtransform/DockerfileBuildingVisitor.java]: Maybe update the model's port aswell?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/visitor/imgtransform/DockerfileBuildingVisitor.java [hunk 5]
```diff
@@ -154,7 +206,11 @@ private void copyAndExecIfPresent(String nodeName, String opName, Optional<Opera
                 }
                 if (path.endsWith(".sql")) return;
                 builder.copyFromCsar(path, nodeName, nodeName + "-" + opName);
-                builder.run("sh " + nodeName + "-" + opName);
+                if (!isStartup) {
+                    builder.run("sh " + nodeName + "-" + opName);
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/visitor/imgtransform/DockerfileBuildingVisitor.java]: I see some duplication here

File: server/src/main/resources/plugins/scripts/util/k8s-cluster-status-check.sh [hunk 6]
```diff
@@ -0,0 +1,6 @@
+function check_k8s_cluster_status () {
+  printf "Checking Kubernetes Cluster Status..."
+
```
Comment (hnicke) [inline server/src/main/resources/plugins/scripts/util/k8s-cluster-status-check.sh]: Did you forget something here?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java (hunk 2 above)
Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java]: if you want approx. 200 Lines saying downloading/extracting/Checking for sure :)

File: server/src/main/resources/plugins/scripts/util/k8s-cluster-status-check.sh (hunk 6 above)
Comment (c-mueller) [reply server/src/main/resources/plugins/scripts/util/k8s-cluster-status-check.sh]: Will do this in a new issue

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java (hunk 2 above)
Comment (hnicke) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java]: This is currently set to `trace`, not `info`.

File: server/src/main/resources/plugins/scripts/util/k8s-cluster-status-check.sh (hunk 6 above)
Comment (hnicke) [reply server/src/main/resources/plugins/scripts/util/k8s-cluster-status-check.sh]: Please add a todo mark.

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java (hunk 2 above)
Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java]: Hmm thats weird. Why is it producing less log output then?

Before: https://gist.github.com/c-mueller/031863766bfbe65b172243eb4bbb169f
After: https://gist.github.com/c-mueller/c1d5d034163b6b488c6b7785f4c47aac

Comment (hnicke) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java]: Maybe its a bug in the docker library.

Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/image/ImageBuilder.java]: Where is the bug? i am changing the level from info to trace. Trace beeing even higher then debug.

**Pull request merged**